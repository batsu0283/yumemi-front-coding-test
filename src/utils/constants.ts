export const populationCategories = [
  {
    label: "総人口",
    value: "total",
  },
  {
    label: "年少人口",
    value: "young",
  },
  {
    label: "生産年齢人口",
    value: "workingAge",
  },
  {
    label: "老年人口",
    value: "elderly",
  },
] as const;

export const defaultHeaders = {
  "X-API-KEY": import.meta.env.VITE_RESAS_API_KEY as string,
} as const;

export const prefectureColors = {
  1: "#FF6633", // 北海道
  2: "#FFB399", // 青森県
  3: "#FF33FF", // 岩手県
  4: "#FFFF99", // 宮城県
  5: "#00B3E6", // 秋田県
  6: "#E6B333", // 山形県
  7: "#3366E6", // 福島県
  8: "#999966", // 茨城県
  9: "#99FF99", // 栃木県
  10: "#B34D4D", // 群馬県
  11: "#80B300", // 埼玉県
  12: "#809900", // 千葉県
  13: "#E6B3B3", // 東京都
  14: "#6680B3", // 神奈川県
  15: "#66991A", // 新潟県
  16: "#FF99E6", // 富山県
  17: "#CCFF1A", // 石川県
  18: "#FF1A66", // 福井県
  19: "#E6331A", // 山梨県
  20: "#33FFCC", // 長野県
  21: "#66994D", // 岐阜県
  22: "#B366CC", // 静岡県
  23: "#4D8000", // 愛知県
  24: "#B33300", // 三重県
  25: "#CC80CC", // 滋賀県
  26: "#66664D", // 京都府
  27: "#991AFF", // 大阪府
  28: "#E666FF", // 兵庫県
  29: "#4DB3FF", // 奈良県
  30: "#1AB399", // 和歌山県
  31: "#E666B3", // 鳥取県
  32: "#33991A", // 島根県
  33: "#CC9999", // 岡山県
  34: "#B3B31A", // 広島県
  35: "#00E680", // 山口県
  36: "#4D8066", // 徳島県
  37: "#809980", // 香川県
  38: "#E6FF80", // 愛媛県
  39: "#1AFF33", // 高知県
  40: "#999933", // 福岡県
  41: "#FF3380", // 佐賀県
  42: "#CCCC00", // 長崎県
  43: "#66E64D", // 熊本県
  44: "#4D80CC", // 大分県
  45: "#9900B3", // 宮崎県
  46: "#E64D66", // 鹿児島県
  47: "#4DB380", // 沖縄県
} as const;

export type PrefectureColors = typeof prefectureColors;
export type PrefCode = keyof PrefectureColors;
export type PopulationCategoriesValue =
  (typeof populationCategories)[number]["value"];
