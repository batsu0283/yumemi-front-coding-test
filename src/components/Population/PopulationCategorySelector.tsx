import { populationCategories } from "../../utils/constants";

type PopulationCategorySelectorProps = {
  selectedPopulationCategory: string;
  handlePopulationCategoryChange: (value: string) => void;
};

export const PopulationCategorySelector = ({
  selectedPopulationCategory,
  handlePopulationCategoryChange,
}: PopulationCategorySelectorProps) => (
  <label className="block text-gray-700 mb-2">
    表示するデータを選択
    <select
      value={selectedPopulationCategory}
      onChange={(e) => handlePopulationCategoryChange(e.target.value)}
      className="block w-full max-w-xs p-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-cyan-500 focus:border-cyan-500"
    >
      {populationCategories.map((populationCategory) => (
        <option key={populationCategory.value} value={populationCategory.value}>
          {populationCategory.label}
        </option>
      ))}
    </select>
  </label>
);
