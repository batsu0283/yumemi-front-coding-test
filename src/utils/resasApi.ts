import { type PrefCode, defaultHeaders } from "./constants";

export type Prefecture = {
  prefCode: PrefCode;
  prefName: string;
};

export type PrefectureAPIResponce = {
  message: string | null;
  result: Prefecture[];
};

export type YearlyPopulation = {
  year: number;
  value: number;
};
export type PopulationData = {
  label: "総人口" | "年少人口" | "生産年齢人口" | "老年人口";
  data: YearlyPopulation[];
};
export type PopulationCompositionPerYearAPIResponce = {
  message: string | null;
  result: {
    data: PopulationData[];
  };
};

const execApi = async <T extends { result?: unknown }>(
  endpoint: string,
): Promise<T> => {
  const url = `https://opendata.resas-portal.go.jp/${endpoint}`;
  const response = await fetch(url, {
    headers: defaultHeaders,
  });

  const data: T = await response.json();
  if (!response.ok || !data.result) {
    const message = `HTTP error! status: ${response.status} url: ${url}`;
    console.error(message);
    throw new Error(message);
  }
  return data;
};

export const resasApi = {
  getPrefectures: async () => {
    const res = await execApi<PrefectureAPIResponce>("api/v1/prefectures");
    return res;
  },

  getPopulationCompositionPerYear: async (queryString: string) => {
    const res = await execApi<PopulationCompositionPerYearAPIResponce>(
      `api/v1/population/composition/perYear?${queryString}`,
    );
    return res;
  },
};
