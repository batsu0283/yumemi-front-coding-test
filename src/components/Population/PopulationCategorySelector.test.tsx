import { fireEvent, render, screen } from "@testing-library/react";
import { populationCategories } from "../../utils/constants";
import { PopulationCategorySelector } from "./PopulationCategorySelector";

describe("PopulationCategorySelector", () => {
  const mockHandlePopulationCategoryChange = vi.fn();
  beforeEach(() => {
    mockHandlePopulationCategoryChange.mockClear();
  });

  it("populationCategoriesがそれぞれ正しくレンダリングされる", () => {
    render(
      <PopulationCategorySelector
        selectedPopulationCategory=""
        handlePopulationCategoryChange={mockHandlePopulationCategoryChange}
      />,
    );

    const selectElement = screen.getByRole("combobox");
    expect(selectElement).toBeInTheDocument();

    for (const category of populationCategories) {
      expect(screen.getByText(category.label)).toBeInTheDocument();
    }
  });

  it("オプションが選択されたときにhandlePopulationCategoryChangeが呼び出されること", () => {
    render(
      <PopulationCategorySelector
        selectedPopulationCategory=""
        handlePopulationCategoryChange={mockHandlePopulationCategoryChange}
      />,
    );

    const selectElement = screen.getByRole("combobox");
    fireEvent.change(selectElement, {
      target: { value: populationCategories[1].value },
    });

    expect(mockHandlePopulationCategoryChange).toHaveBeenCalledTimes(1);
    expect(mockHandlePopulationCategoryChange).toHaveBeenCalledWith(
      populationCategories[1].value,
    );
  });

  it("Snapshot", () => {
    const { container } = render(
      <PopulationCategorySelector
        selectedPopulationCategory=""
        handlePopulationCategoryChange={mockHandlePopulationCategoryChange}
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
