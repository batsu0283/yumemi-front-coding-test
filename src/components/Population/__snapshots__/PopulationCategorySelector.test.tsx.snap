// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PopulationCategorySelector > matches snapshot 1`] = `
<div>
  <label
    class="block text-gray-700 mb-2"
  >
    表示するデータを選択
    <select
      class="block w-full max-w-xs p-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-cyan-500 focus:border-cyan-500"
    >
      <option
        value="total"
      >
        総人口
      </option>
      <option
        value="young"
      >
        年少人口
      </option>
      <option
        value="workingAge"
      >
        生産年齢人口
      </option>
      <option
        value="elderly"
      >
        老年人口
      </option>
    </select>
  </label>
</div>
`;

exports[`matches snapshot 1`] = `
<div>
  <label
    class="block text-gray-700 mb-2"
  >
    表示するデータを選択
    <select
      class="block w-full max-w-xs p-2 border border-gray-300 rounded-md shadow-sm focus:outline-none focus:ring-cyan-500 focus:border-cyan-500"
    >
      <option
        value="total"
      >
        総人口
      </option>
      <option
        value="young"
      >
        年少人口
      </option>
      <option
        value="workingAge"
      >
        生産年齢人口
      </option>
      <option
        value="elderly"
      >
        老年人口
      </option>
    </select>
  </label>
</div>
`;
